File: pynng/reqrep_time_protobuf/requester.py
# Copyright (C) 2023 twyleg
import time_pb2
import asyncio
import pynng


address = "ipc:///tmp/reqrep_time_date_protobuf.ipc"


async def request_time() -> None:
    with pynng.Req0(dial=address) as req:

        while True:
            print("Requesting time!")
            request = time_pb2.Request()
            request.request_type = time_pb2.Request.RequestType.REQUEST_TYPE_TIME
            req.send(request.SerializeToString())

            response = req.recv_msg()
            time = time_pb2.Time()
            time.ParseFromString(response.bytes)

            print(f"Received time: {time}")
            await asyncio.sleep(1.0)


async def request_date() -> None:
    with pynng.Req0(dial=address) as req:

        while True:
            print("Requesting date!")
            request = time_pb2.Request()
            request.request_type = time_pb2.Request.RequestType.REQUEST_TYPE_DATE

            req.send(request.SerializeToString())

            response = req.recv_msg()
            date = time_pb2.Time()
            date.ParseFromString(response.bytes)
            print(f"Received date: {date}")
            await asyncio.sleep(5.0)


if __name__ == '__main__':

    loop = asyncio.get_event_loop()
    request_time_task = loop.create_task(request_time())
    request_date_task = loop.create_task(request_date())
    loop.run_until_complete(asyncio.wait([request_time_task, request_date_task]))

    loop.run_forever()

File: nngpp/reqrep_time_protobuf/requester/requester.cc
// Copyright (C) 2023 twyleg
#include <time.pb.h>

#include <nngpp/nngpp.h>
#include <nngpp/protocol/req0.h>
#include <nngpp/platform/platform.h>

#include <string>
#include <chrono>
#include <thread>
#include <iostream>
#include <iomanip>


using namespace std::literals::chrono_literals;
using namespace std::literals::string_literals;


const std::string address = "ipc:///tmp/reqrep_time_date_protobuf.ipc";


void request(playground::nng::time::Request::RequestType requestType, std::chrono::milliseconds delay) {

	auto req = nng::req::open();
	req.dial(address.c_str());

	while(true) {
		std::cout << "Requesting " << requestType << std::endl;

		playground::nng::time::Request request;
		request.set_request_type(requestType);

		const auto serializedObject = request.SerializeAsString();

		req.send(nng::view(serializedObject.data(), serializedObject.size()));

		auto msg = req.recv_msg();
		const nng::view msgView = msg.body().get();

		switch(requestType) {
		case playground::nng::time::Request::REQUEST_TYPE_TIME: {

			playground::nng::time::Time time;
			time.ParseFromArray(msgView.data(), msgView.size());
			std::cout << "Received time:" << time.DebugString() << std::endl;
			break;
		}
		case playground::nng::time::Request::REQUEST_TYPE_DATE: {

			playground::nng::time::Date date;
			date.ParseFromArray(msgView.data(), msgView.size());
			std::cout << "Received date:" << date.DebugString() << std::endl;
			break;
		}
		}

		std::this_thread::sleep_for(delay);
	}

}


int main(int argc, char *argv[]) {
	std::cout << "Started reqrep_time_request\n\r";
	std::thread requestTimeThread(request, playground::nng::time::Request::REQUEST_TYPE_TIME, 1000ms);
	std::thread requestDateThread(request, playground::nng::time::Request::REQUEST_TYPE_DATE, 5000ms);

	requestTimeThread.join();
	requestDateThread.join();

	return 0;
}
